pyirf/interpolation.py

```diff
@@ -2,22 +2,239 @@
 
 import numpy as np
 import astropy.units as u
-from scipy.interpolate import griddata
+from scipy.interpolate import griddata, interp1d
+from scipy.spatial import Delaunay
 from pyirf.utils import cone_solid_angle
+from pyirf.binning import bin_center
+
+import warnings
 
 __all__ = [
-    'interpolate_effective_area_per_energy_and_fov',
-    'interpolate_energy_dispersion',
+    "interpolate_effective_area_per_energy_and_fov",
+    "interpolate_energy_dispersion",
 ]
 
 
-@u.quantity_input(effective_area=u.m**2)
+def lookup(x, len_hist):
+    """
+    Function to look up the bin-value corresponding to a bin number.
+
+    Parameters
+    ----------
+    x: numpy.ndarray, shape=(M+L)
+    Concatenated array in which the first M values are the M bin-values for a histogram consisting of M bins and
+    the following L values are L bin-numbers, where the histogram value should be looked up.
+
+    len_hist: int
+    Number (M) of histogram bins
+
+    Returns
+    -------
+    looked_up_values: numpy.ndarray, shape=(M)
+    Looked up bin-values
+    """
+    hist = x[:len_hist]
+    binnr = x[len_hist:]
+
+    mask = (binnr == 0) | (binnr == len_hist + 1)
+
+    looked_up_values = np.zeros(len(x) - len_hist)
+    looked_up_values[~mask] = hist[(binnr[~mask] - 1).astype(int)]
+
+    return looked_up_values
+
+
+def numerical_quantile(pdf, mids, percentile):
+    """
+    Approximative quantile computation for histograms
+
+    Parameters
+    ----------
+    pdf: numpy.ndarray, shape=(2)
+    Input histogram
+
+    mids: numpy.ndarray, shape=(M)
+    Bin mids for the input histogram
+
+    percentile: numpy.ndarray, shape=(L)
+    Percentages where the quantile is needed
+
+    Returns
+    -------
+    quantiles: numpy.ndarray, shape=(L)
+    Computed quantiles
+
+    """
+    if np.sum(pdf) == 0:
+        return np.full(len(percentile), np.nan)
+
+    # Compute cdf and interpolate for ppf, Normalize to one first, since the sum of a histogram does not
+    # need to be one
+    cdf = np.cumsum(pdf)
+    interp = interp1d(
+        cdf / cdf.max(), mids, fill_value=(mids.min(), mids.max()), bounds_error=False
+    )
+
+    # Lookup ppf value
+    return interp(percentile)
+
+
+def rebin(entries, mids, width):
```

Comment:
An elegant solution for this problem is to use https://numpy.org/doc/stable/reference/generated/numpy.ufunc.reduceat.html, this way you get ND support and can avoid the Python for loop. However for rebinning along a single axis, the Python for loop might be performant enough. You can use it with arbitrary bins size, just defined by the bin indices.